#1089: IDE2 falls back to new sketch if opening failed.

kittaakos opened the pull request:
### Motivation
<!-- Why this pull request? -->

IDE2 must automatically open a new sketch (in the temp folder) if the sketch was deleted between the IDE2 stop/start cycle.

Requirements:
 - At least three valid sketches in the sketchbook folder. My bare minimum setup is the following:
```
a.kitta@Akoss-MacBook-Pro Arduino % ls -al
total 16
drwxr-xr-x  6 a.kitta  staff   192 Jul 14 15:03 .
drwx------+ 5 a.kitta  staff   160 Jul  6 15:46 ..
-rw-r--r--@ 1 a.kitta  staff  6148 Jul 14 15:03 .DS_Store
drwxr-xr-x  3 a.kitta  staff    96 Jul  7 16:42 sketch_1
drwxr-xr-x  3 a.kitta  staff    96 Jul  7 16:43 sketch_2
drwxr-xr-x  3 a.kitta  staff    96 Jul  7 16:43 sketch_3
a.kitta@Akoss-MacBook-Pro Arduino % cat sketch_1/sketch_1.ino 
void setup() {}
void loop() {}
```

(`sketch_2` and `sketch_3` have the same content as `sketch_1`).

Preconditions: 
 - Open IDE2, if `sketch_1` is not loaded, open `sketch_1`,
 - Close any other opened sketches
 - When `sketch_1` is the only opened sketch in IDE2 the verification can start.

1: Opening deleted sketch from `File` > `Sketchbook`
 - Note: the file-system watcher might miss file deletion events and cannot update the `Sketchbook` submenu content.
 - Check `File` > `Sketchbook`, `sketch_3` is there,
 - Delete `sketch_3` outside from IDE2,
 - Open `File` > `Sketchbook` > `sketch_3`,
 - See error notification that it's missing.
 - Check `File` > `Sketchbook`, `sketch_3` is not there.

https://user-images.githubusercontent.com/1405703/179183140-c5018ca3-77e8-4785-af61-01674c718677.mp4

----

2. Delete the sketch folder when IDE2 is stopped:
 - Open `sketch_1` in IDE2,
 - Close IDE2,
 - Delete `sketch_1` from the sketchbook,
 - Start IDE2,
 - Without any error notifications, IDE2 opens a new fallback sketch


https://user-images.githubusercontent.com/1405703/179183207-02ca16e1-1741-477e-8219-e2a1e23dd6bc.mp4


----
3. Delete the main sketch file when IDE2 is stopped:
 - Open `sketch_2` in IDE2,
 - Close IDE2,
 - Delete the main sketch file `sketch_2/sketch_2.ino`,
 - Start the IDE2,
 - IDE2 loads, expect a reload,
 - New sketck is created and opened but no error notification is shown by IDE2.


https://user-images.githubusercontent.com/1405703/179183362-2521a55a-675d-46f9-9a3c-78301cf26523.mp4


### Change description
<!-- What does your code do? -->

### Other information

Closes #1089

Breaking: _Optimize for Debugging_ now has its preference: `arduino.compile.optimizeForDebug`. The default value is `false`. If you had _Optimize for Debugging_ enabled before this IDE2 version, you must click on _Optimized for Debugging_ again.

<img width="260" alt="Screen Shot 2022-07-08 at 11 05 51" src="https://user-images.githubusercontent.com/1405703/177958260-26104c60-3931-4c34-9bc4-208bb7c661d7.png">

Breaking: Changed `arduino.cloud.sketchSyncEnpoint` to `arduino.cloud.sketchSyncEndpoint` (#1067). If you had a custom sketch sync endpoint defined, you must redefine it again with the version of the IDE2.

IDE2 uses the default spinner from Theia:

https://user-images.githubusercontent.com/1405703/178996866-ce705e24-7245-4cee-859f-5ce717ded231.mp4


### Reviewer checklist

* [ ] PR addresses a single concern.
* [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-ide/pulls) before creating one)
* [ ] PR title and description are properly filled.
* [ ] Docs have been added / updated (for bug fixes / features)

kittaakos commented on arduino-ide-extension/src/node/sketches-service-impl.ts:
```diff
@@ -589,6 +584,14 @@ interface SketchWithDetails extends Sketch {
+  return (
+    ServiceError.is(err) &&
+    err.code === 5 &&
+    err.message.toLocaleLowerCase('en-US').includes('not_found')
```
Probably superfluous: https://grpc.github.io/grpc/core/md_doc_statuscodes.html.

kittaakos commented on arduino-ide-extension/src/common/protocol/sketches-service.ts:
```diff
@@ -1,5 +1,21 @@
+    (message: string, uri: string) => {
+      return {
+        message,
+        data: { uri },
```
Consider including the file-system path? Would be obsolete with the next Theia version.

kittaakos replied:
350c04b0bc28bc57acb12d65c6b12cff1777ccea + https://github.com/arduino/arduino-cli/issues/1797

kittaakos commented on arduino-ide-extension/src/common/protocol/sketches-service-client-impl.ts:
```diff
@@ -110,9 +117,12 @@ export class SketchesServiceClientImpl
+      this._currentSketch = currentSketch;
+      this.currentSketchDidChangeEmitter.fire(this._currentSketch);
+      this.currentSketchLoaded.resolve(this._currentSketch);
+    }, 1_000);
```
on app `ready`?

kittaakos replied:
ee17fd4

per1234 reviewed (changes requested):
There are two distinct reasons the IDE might attempt to open a sketch:

- Explicit action by user (e.g., selecting a sketch from **File > Sketchbook**)
- Restore state from previous session (https://github.com/arduino/arduino-ide/pull/780)

The IDE must gracefully handle a failure of either operation (https://github.com/arduino/arduino-ide/issues/1089). However, I feel the need to communicate about the failure to the user differs between them.

If an operation triggered explicitly by the user fails, it is important to communicate this to the user. So I think showing a notification when the sketch opening fails is necessary in this case.

The situation is different when it comes to automatically opening sketches from the previous IDE session. This is only a convenience feature. If it is able to open the sketches, great. But if it is not able to open the sketches, I don't want to hear about it. I might have intentionally deleted that sketch. It might have been a new sketch that was staged in a temporary folder the OS cleaned up. In the latter case, the temporary path makes the error message especially cryptic:
![image](https://user-images.githubusercontent.com/8572152/179133865-07c33cb4-1485-4d02-806d-1c57acb98af6.png)

So I suggest that no error notification be shown when the sketch opening fails during the restore operation (logging it is fine of course).

kittaakos commented:
Thank you for the review!

> So I suggest that no error notification be shown when the sketch opening fails during the restore operation (logging it is fine of course).

Sure 👍 

IDE2 (and the CLI) might fail to load the sketch due to two reasons:
 - the sketch folder was deleted,
 - the sketch is not valid (the main file was deleted, does not match the folder name, .etc)

As a user, do you expect that IDE2 will omit the error notification in both cases?

per1234 commented:
>do you expect that IDE2 will omit the error notification in both cases?

In the case of a "restore" operation, yes.

kittaakos commented:
I have updated the PR not to show the error notification when restoring the previously opened sketch has failed (due to a missing or invalid sketch). I updated the screencasts too to show the expected IDE2 behavior. Please review.

per1234 reviewed (approved):
It works perfectly. Excellent work Akos!

AlbyIanna commented on arduino-ide-extension/src/browser/contributions/first-start-installer.ts:
```diff
@@ -0,0 +1,45 @@
+const INIT_LIBS_AND_PACKAGES = 'initializedLibsAndPackages';
+
+@injectable()
+export class FirstStartInstaller extends Contribution {
+  @inject(LocalStorageService)
+  private readonly localStorageService: LocalStorageService;
+
+  @inject(BoardsService)
+  private readonly boardsService: BoardsService;
+
+  @inject(LibraryService)
+  private readonly libraryService: LibraryService;
+
+  override onReady(): void {
+    this.localStorageService
+      .getData(INIT_LIBS_AND_PACKAGES)
+      .then(async (value) => {
+        const isFirstStartup = !value;
+        if (isFirstStartup) {
+          await this.localStorageService.setData(INIT_LIBS_AND_PACKAGES, true);
+          const avrPackage = await this.boardsService.getBoardPackage({
+            id: 'arduino:avr',
+          });
+          const builtInLibrary = (
+            await this.libraryService.search({
+              query: 'Arduino_BuiltIn',
+            })
+          )[0];
+
+          !!avrPackage &&
+            (await this.boardsService.install({ item: avrPackage }));
+          !!builtInLibrary &&
+            (await this.libraryService.install({
+              item: builtInLibrary,
+              installDependencies: true,
+            }));
+        }
+      });
+  }
+}
```
Make sure to remove this as we already put this code [into a separate contribution](https://github.com/arduino/arduino-ide/commit/73835eced3bda7657308bf582bd5e5690d73879a#diff-8e1a8da61c276800216f425dc94832341b6b2b17639b36e837399fb13102b128).

AlbyIanna reviewed (approved):
LGTM.

I left some comments, but nothing important.

kittaakos commented on arduino-ide-extension/src/browser/boards/boards-service-provider.ts:
```diff
@@ -39,6 +39,9 @@ export class BoardsServiceProvider implements FrontendApplicationContribution {
   @inject(NotificationCenter)
   protected notificationCenter: NotificationCenter;
 
+  @inject(FrontendApplicationStateService)
+  private readonly appStateService: FrontendApplicationStateService;
+
```
Good idea, but I do not want to introduce a hard dependency between contributions. Also, this service does not contribute any commands, menu items, .etc.
